### PR TITLE
[e2e tests] Fix flaky create grouped product test

### LINKED ISSUE #50820: [Flaky Test] can create a grouped product
**Flaky test detected. This is an auto-generated issue by GitHub Actions. Please do NOT edit this manually.**

## Test title
can create a grouped product

## Test path
`/plugins/woocommerce/tests/e2e-pw/tests/product/block-editor/create-grouped-product-block-editor.spec.js`

## Errors

<time datetime="2024-08-21T00:15:33.109Z"><code>[2024-08-21T00:15:33.109Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/10480870733"><code>trunk</code></a>.
<br/>
<time datetime="2024-10-09T13:45:22.937Z"><code>[2024-10-09T13:45:22.937Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11255940925"><code>test/ts</code></a>.
<br/>
<time datetime="2024-10-29T15:58:51.348Z"><code>[2024-10-29T15:58:51.348Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11577428332"><code>enhance/remote-logging-data-privacy</code></a>.
<br/>
<time datetime="2024-11-05T11:57:41.479Z"><code>[2024-11-05T11:57:41.479Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11683330399"><code>add/status-indicator-component</code></a>.
<br/>
<time datetime="2024-11-06T14:59:25.256Z"><code>[2024-11-06T14:59:25.256Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11705826553"><code>52217-improve-email-logo-settings</code></a>.
<br/>
<time datetime="2024-11-15T15:38:48.081Z"><code>[2024-11-15T15:38:48.081Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11858781339"><code>add/checkout-field-entities-42773</code></a>.
<br/>
<time datetime="2024-11-15T15:55:45.126Z"><code>[2024-11-15T15:55:45.126Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11858781339"><code>add/checkout-field-entities-42773</code></a>.
<br/>
<time datetime="2024-11-21T23:13:52.588Z"><code>[2024-11-21T23:13:52.588Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/11963002016"><code>fix/43256-tablecard-columns-popover</code></a>.
<br/>
<time datetime="2024-11-29T11:22:59.568Z"><code>[2024-11-29T11:22:59.568Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/12083557158"><code>add/categories-to-setting-page</code></a>.
<br/>
<time datetime="2024-12-02T11:43:20.709Z"><code>[2024-12-02T11:43:20.709Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/12118305262"><code>dev/nox-drag-and-drop-updates</code></a>.
<br/>
<time datetime="2025-04-03T15:58:15.361Z"><code>[2025-04-03T15:58:15.361Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14246913690"><code>test/performance-patch</code></a>.
<br/>
<time datetime="2025-04-03T16:06:42.792Z"><code>[2025-04-03T16:06:42.792Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14246877582"><code>wooplug-2760-product-collection-all-templates-page-performance</code></a>.
<br/>
<time datetime="2025-04-03T16:37:31.561Z"><code>[2025-04-03T16:37:31.561Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14247711907"><code>test/performance-patch</code></a>.
<br/>
<time datetime="2025-04-03T17:09:09.486Z"><code>[2025-04-03T17:09:09.486Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14248343744"><code>test/performance-patch</code></a>.
<br/>
<time datetime="2025-04-16T19:16:59.050Z"><code>[2025-04-16T19:16:59.050Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14500483806"><code>56715-implement-a-way-for-developers-to-register-an-address-autocomplete-provider-in-php</code></a>.
<br/>
<time datetime="2025-04-16T19:34:21.023Z"><code>[2025-04-16T19:34:21.023Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/14500483806"><code>56715-implement-a-way-for-developers-to-register-an-address-autocomplete-provider-in-php</code></a>.
<br/>
<details>
<summary>
	<time datetime="2025-08-09T03:44:11.500Z"><code>[2025-08-09T03:44:11.500Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/woocommerce/woocommerce/actions/runs/16844893273"><code>trunk</code></a>.
</summary>

```
Error: locator.click: Error: strict mode violation: getByText('Small Metal Fish') resolved to 2 elements:
    1) <div class="woocommerce-add-products-modal__menu-item-title">Small Metal Fish</div> aka locator('#woocommerce-experimental-select-control-0-item-1').getByText('Small Metal Fish')
    2) <div class="woocommerce-add-products-modal__menu-item-title">Small Metal Fish</div> aka locator('#woocommerce-experimental-select-control-0-item-2').getByText('Small Metal Fish')

Call log:
  - waiting for getByText('Small Metal Fish')

    at /home/runner/work/woocommerce/woocommerce/plugins/woocommerce/tests/e2e-pw/tests/product/block-editor/create-grouped-product-block-editor.spec.js:109:42
```
</details>

### PR DESCRIPTION
**Pull request by veljkho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #50820

These tests were pretty non flaky but sometimes flaky. As Buildkite says, over 98% reliability score.

The flaky part was waiting for a response. I added a regression step instead of waiting for a response to fix the flakiness and increase coverage. It worked well locally, both with and without Gutenberg—100% success.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

```
pnpm test:e2e tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js --repeat-each=10
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js [hunk 1]
```diff
@@ -72,11 +72,9 @@ test.describe( 'General tab', { tag: '@gutenberg' }, () => {
...
 				.click();
 
-			await page.waitForResponse(
-				( response ) =>
-					response.url().includes( '/wp-json/wc/v3/products/' ) &&
-					response.status() === 200
-			);
+			await expect(
+				page.getByLabel( 'Dismiss this notice' )
+			).toContainText( 'Product type changed.' );
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: When we switch the product type, in this case to a grouped type, there's a saving in progress (draft). I added waiting for the notice of successful saving and then proceeding further.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js [hunk 2]
```diff
@@ -86,15 +86,19 @@ test.describe( 'General tab', { tag: '@gutenberg' }, () => {
...
+					.type( categoryName, { delay: 100 } );
+				await page
+					.getByRole( 'option', { name: categoryName } )
+					.locator( 'label' )
+					.waitFor();
+				await expect(
+					page
+						.getByRole( 'option', { name: categoryName } )
+						.locator( 'label' )
+				).toHaveText( categoryName );
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: There's probably some background processes in combination to a very fast Playwright that hides the category from showing up in the results at first point, so I added action to click and type the category name just like a normal end user would type, which adds some time and shows the category in the end. This way we also cover typing in the category field and searching for a category, which is nice regression to have.

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: I'm don't think removing `waitForCategoriesResponse` is a good idea. This is a very dynamic area and I remember having a lot of trouble stabilizing it and waitForCategoriesResponse fixed it. I see no harm in leaving it. And I'm very sure it's not a cause of flakiness, so why remove it?

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: > I added action to click and type the category name just like a normal end user would type, which adds some time and shows the category in the end.

Waiting for 100ms does nothing to guarantee the categories are loaded.
The `waitForCategoriesResponse` is there to make sure the categories are fetched before typing. If they are not, the test will fail.

Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: I understand the reasoning behind keeping `waitForCategoriesResponse` — it worked in the past to stabilize the test, and I can see how that was necessary given the dynamic nature of this area. However, I encountered a specific issue where the test failed due to the timing of the response, and I believe waiting for the response alone wasn’t enough to guarantee the categories were loaded in time.

The test was failing intermittently while waiting for the response from waitForCategoriesResponse, and this was the root cause of the issue. The delay of 100ms between keystrokes I added to simulate typing was working well in stabilizing the input, but it wasn’t enough to handle the timing of the response in some cases.

As a result, I added a more explicit wait to ensure that the category option was not only typed into the field but also loaded and visible before interacting with it. The combination of waiting for the option’s label to appear and asserting the text seemed to be an effective way to prevent flaky failures, and I’ve executed it locally several times with consistent success.

I think the solution I implemented is more robust because it:

1. Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.
2. Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.

While `waitForCategoriesResponse` helped us in the past, I believe this additional wait for the category option’s label ensures better synchronization, reducing the chances of race conditions and preventing test failures.

I’m happy to discuss further if needed and hear from others as well.

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: > I understand the reasoning behind keeping `waitForCategoriesResponse` — it worked in the past to stabilize the test, and I can see how that was necessary given the dynamic nature of this area. However, I encountered a specific issue where the test failed due to the timing of the response, and I believe waiting for the response alone wasn’t enough to guarantee the categories were loaded in time.
> 
> The test was failing intermittently while waiting for the response from waitForCategoriesResponse, and this was the root cause of the issue. The delay of 100ms between keystrokes I added to simulate typing was working well in stabilizing the input, but it wasn’t enough to handle the timing of the response in some cases.
> 
> As a result, I added a more explicit wait to ensure that the category option was not only typed into the field but also loaded and visible before interacting with it. The combination of waiting for the option’s label to appear and asserting the text seemed to be an effective way to prevent flaky failures, and I’ve executed it locally several times with consistent success.
> 
> I think the solution I implemented is more robust because it:
> 
> 1. Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.
> 2. Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.
> 
> While `waitForCategoriesResponse` helped us in the past, I believe this additional wait for the category option’s label ensures better synchronization, reducing the chances of race conditions and preventing test failures.
> 
> I’m happy to discuss further if needed and hear from others as well.

Can you point me to at least one failure on this `waitForCategoriesResponse`? From what I could find there was a single failure (the last one) waiting for a response, but it was for the `/wp-json/wc/v3/products/`, at line 75, not this one. And the previous failures were for different causes. Please don't simply remove working code by guessing.

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: > * Ensures categories are fully loaded before typing by waiting for the category option’s label to appear.

I'm sorry but this is not how it works. If you start typing before the categories are fetched it will fail. Typing will display the list of categories, and if the fetch finishes after the typing starts the list is not updated and the category will not be found.

> * Avoids unnecessary waits — it doesn’t just blindly wait for the response but explicitly ensures the option is available before proceeding.

This is not how waiting for the response works. There's nothing blind about it.

**Review by adimoldovan — CHANGES_REQUESTED** [review]
@veljkho Please revisit my comments above. Not only these updates are not related to the flakiness, but they will make this test even more unstable.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js (hunk 2 above)
Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: Uhm, this test `linked-product-tab-product-block-editor.spec.js` is not even reported as flaky. Why would you "fix" it? 
> The test was failing intermittently while waiting for the response from waitForCategoriesResponse

Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: @adimoldovan , I have nothing against your implementation. I found it flaky and [here for example](https://github.com/woocommerce/woocommerce/actions/runs/12068850803/job/33655488067) it failed twice in a row. However, since we are dealing with such a low flakiness score and it is working well for the last many runs, let's just keep it as is. We could even combine both implementations if you'd like.

Please give it another look and let me know if I can merge it.

**Comment by veljkho** [comment]
Hi @adimoldovan 

Thanks for your feedback! I appreciate your perspective and understand your concerns about potentially introducing instability to the test. Let me clarify my reasoning further and address your points:

Not Related to Flakiness:

- The primary reason for my change was to address failures I observed related to timing, specifically when the categories were not fully loaded before the test attempted further actions. While `waitForCategoriesResponse` ensures the API request is completed, it doesn’t fully guarantee that the categories are rendered and interactable on the page.

- By explicitly waiting for the category option label to appear and verifying its content, I aimed to provide an additional layer of certainty that the UI is ready before interacting with it.

Concerns About Instability:

- I’d like to better understand how you feel these changes could make the test more unstable. From my testing, waiting for the DOM element directly has been more reliable, as it aligns the test flow with the actual state of the UI. If there are specific scenarios where this might cause instability, I’m open to discussing and finding a middle ground.

- My changes worked consistently in local tests and may add more resilience, especially in environments where UI rendering might lag behind the API response.

Proposed Next Steps:
- If we reencounter flakiness in the future, primarily related to timing or UI readiness, I suggest we combine both approaches. For now, let's use only `waitForCategoriesResponse`.

Ultimately, my goal is to ensure the test is as robust and stable as possible. I’m open to iterating on this further if needed.

**Comment by adimoldovan** [comment]
> Hi @adimoldovan
> 
> Thanks for your feedback! I appreciate your perspective and understand your concerns about potentially introducing instability to the test. Let me clarify my reasoning further and address your points:
> 
> Not Related to Flakiness:
> 
> * The primary reason for my change was to address failures I observed related to timing, specifically when the categories were not fully loaded before the test attempted further actions. While `waitForCategoriesResponse` ensures the API request is completed, it doesn’t fully guarantee that the categories are rendered and interactable on the page.
> * By explicitly waiting for the category option label to appear and verifying its content, I aimed to provide an additional layer of certainty that the UI is ready before interacting with it.
> 
> Concerns About Instability:
> 
> * I’d like to better understand how you feel these changes could make the test more unstable. From my testing, waiting for the DOM element directly has been more reliable, as it aligns the test flow with the actual state of the UI. If there are specific scenarios where this might cause instability, I’m open to discussing and finding a middle ground.
> * My changes worked consistently in local tests and may add more resilience, especially in environments where UI rendering might lag behind the API response.
> 
> Proposed Next Steps:
> 
> * If we reencounter flakiness in the future, primarily related to timing or UI readiness, I suggest we combine both approaches. For now, let's use only `waitForCategoriesResponse`.
> 
> Ultimately, my goal is to ensure the test is as robust and stable as possible. I’m open to iterating on this further if needed.

@veljkho Replying with very long AI generated responses is a waste of both mine and your time. Please review my comments and the code I'm addressing. Everything in this reply is already addressed with clarity. Please stop doing this.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js (hunk 2 above)
Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js]: @veljkho 
1. That is a PR in progress. 
2. That is not a flaky occurrence. In that run the test failed - it is broken. Coincidently that is a PR I am working on and that failure was expected in that case.

Do you have a flakiness report on this test?

**Comment by veljkho** [comment]
@adimoldovan  I don't understand why are you behaving like this. Please review the code as I have added new commit. Thanks.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js (hunk 1 above)
Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: I don't agree that this change will fix the flakiness. If the `/wp-json/wc/v3/products/` request fails the notice will not be displayed either. We need to understand why the the `/wp-json/wc/v3/products/`failed first.

**Review by adimoldovan — CHANGES_REQUESTED** [review]
> The flaky part was waiting for a response. I added a regression step instead of waiting for a response to fix the flakiness and increase coverage. 

I'm sorry but I don't agree. `waiting for a response` may be the point of failure, it doesn't mean it is the cause of flakiness. We need to first understand why the response failed. Removing the check for that response will not fix it.


Also see [my code comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1864256758)

**Comment by veljkho** [comment]
Hi @adimoldovan 
I appreciate that you shared your thoughts, but I feel it’s important to address the tone of our conversation. While disagreement is a natural part of collaboration, how it’s communicated can make a big difference. I felt that your responses lacked respectful disagreement and constructive feedback, **which are essential for fostering a productive and collaborative environment**.

On the topic itself, I’d like to reiterate why `waitForResponse` should not replace UI-based assertions in E2E testing:

The primary purpose of e2e testing is to simulate real user interactions and validate that the application behaves as expected from the user’s perspective. This means relying on UI elements for assertions, as they directly reflect the user experience. Replacing `waitForResponse` with assertions based on UI elements ensures that the test reflects the actual experience users would have. This also means if further, the test fails by, e.g., adding products, we will be notified about it from an e2e perspective, and it could be a bug report (theoretically).

Overusing `waitForRespons` ties tests to specific API behaviors, making them fragile and prone to breaking with backend changes or delays, even when the user facing behavior is unaffected.

> Removing the check for that response will not fix it.
> I don't agree that this change will fix the flakiness. If the /wp-json/wc/v3/products/ request fails the notice will not be displayed either. We need to understand why the the /wp-json/wc/v3/products/failed first.

Here is the proof that the notice was indeed present when it failed to wait for a response:
- [**Please check this video**](https://d.pr/i/7uBZ3Z).

Failures in `waitForResponse` can be caused by backend issues, browser network conditions, or other external factors. These may not impact the user experience directly but can still cause the test to fail unnecessarily. By relying on UI checks, the test evaluates the final result that matters to the user, independent of backend instability.

Backend behavior is better validated through API tests. For example, if an API call fails, it should be addressed in a dedicated API test where its behavior can be isolated, rather than in an E2E test that focuses on the user interface.

I hope this summary clarifies my perspective and encourages a more collaborative approach moving forward. To address a prior remark, this long text is not an AI-generated response but **a well-structured and carefully considered perspective aimed at fostering constructive collaboration**.

**Comment by adimoldovan** [comment]
> Hi @adimoldovan I appreciate that you shared your thoughts, but I feel it’s important to address the tone of our conversation. While disagreement is a natural part of collaboration, how it’s communicated can make a big difference. I felt that your responses lacked respectful disagreement and constructive feedback, **which are essential for fostering a productive and collaborative environment**.
> 

@veljkho If you need to take this to a personal level, let's take it offline. My feedback was exactly constructive, on code, with arguments. And respectful. If you felt it was not respectful, I'm sorry about it.

> On the topic itself, I’d like to reiterate why `waitForResponse` should not replace UI-based assertions in E2E testing:
...

I'm not saying it should replace anything. But if you would spend some time to actually understand this flow and to actually read my feedback you would understand that checking for that response is key to the stability of this test.

> Here is the proof that the notice was indeed present when it failed to wait for a response:
[Please check this video]().

Veljko, nowhere in that video you can see the API request failed. You are showing a successful flow. 

> Failures in waitForResponse can be caused by backend issues, browser network conditions, or other external factors. These may not impact the user experience directly but can still cause the test to fail unnecessarily.

Velkjo, if that API request fails it means the product is not updated.

> To address a prior remark, this long text is not an AI-generated response but **a well-structured and carefully considered perspective aimed at fostering constructive collaboration**. 

I would rather stop this here. It's clear to me that you're unwilling to stop with this nonsene.

Please, if you want to collaborate on this PR, let's talk on code.

**Comment by veljkho** [comment]
Hey @adimoldovan 

Thank you for sharing your thoughts. I truly believe we’re both working toward the same goal.

> Veljko, nowhere in that video you can see the API request failed. You are showing a successful flow.

That video is exactly from a failed e2e test where it failed to check `waitForResponse`.

While our perspectives differ on this issue, I’ve made some adjustments to ensure the test is as robust and accurate as possible:
- A UI-based check was added to validate the presence of the notice and confirm that the user-facing behavior is correct.
- Modified the `waitForResponse` part of the test to ensure the API response is received successfully, verify that the backend process functions as expected and avoid race conditions (which seems to be the case here)

By combining these two checks, I believe we can address both user experience and backend functionality concerns, making the test more comprehensive and reliable.

These updates have worked consistently for me locally, and I’m confident this approach aligns with best practices. However, your feedback is essential to ensure we’re on the right track. Could you please review the latest changes and let me know your thoughts?

Looking forward to collaborating further to finalize this test!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js (hunk 1 above)
Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: The notice was indeed present even after failing to verify the `waitForResponse`. I believe this might be a rare race condition, as it happens very infrequently, and the test’s reliability score remains quite high.

To address this, I’ve moved the `waitForResponse` earlier in the process and adjusted the code accordingly. Could you please review the latest changes and let me know your thoughts?

Comment (juliaamosova) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: >To address this, I’ve moved the `waitForResponse` earlier in the process and adjusted the code accordingly.

@veljkho we still haven't figured out why waitForResponse may be failing in its original place (after the user clicks on _Change product type_ and selects _Grouped product_). As @adimoldovan mentioned above, this is a key to understanding why the test is flaky. 

What are the reasons you moved `waitForResponse` after the user clicks on _General_ tab? 

I inspected API calls being made when the user selects _Grouped product_ and I indeed saw the POST request to create the product:

api_response

With that said, we need to keep checking for a successful response in its original place to ensure the product draft gets created before proceeding. 

I further researched how to use `waitForResponse` properly and I came across [this article](https://www.workwithloop.com/blog/our-1-solution-to-playwright-flakiness-waitforresponse-waitforrequest-promises) that suggests using [Promise ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)with `waitForResponse`. If we use `waitForResponse` without a promise - quoting from the article further:
>There is a risk that the API may have already triggered and received the response by the time Playwright starts waiting for it. This can occur if the API is particularly fast, or if there is a delay in the execution of the test script due to other factors. If this happens, the waitForResponse will time out and the test will fail.

I believe this is what happens in our case - judging by the [video of the failed test you shared here](https://github.com/woocommerce/woocommerce/pull/53310#issuecomment-2510936439) and then [mentioned](https://github.com/woocommerce/woocommerce/pull/53310#issuecomment-2511723070) that it was a case when the test actually failed despite the video showing that the flow was successful. 

As a next step here, I would try to use a Promise to make the check for API response to mitigate the risk of the test failing due to flakiness in this place. 

I also see that you added the check for the notice `Product type changed` to appear after the change of product type. As we saw in the video you shared, the notice appeared even after the `waitForResponse` failed so the check for the notice doesn't help here and can be removed.

**Review by juliaamosova — CHANGES_REQUESTED** [review]
@veljkho I left a [comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867092458) related to your recent code changes. Please review and make further changes as discussed in the comment.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js (hunk 1 above)
Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: Thanks @juliaamosova for this constructive feedback that I was looking to receive all the time. I really appreciate the time you put into this.

> this is a key to understanding why the test is flaky.

In my investigation, I already understood that this was due to a race condition and the inappropriate use of `waitForResponse`. Even with the original approach, we are dealing with a very low flakiness report according to BK, where we have currently only 1 instance out of 100 last runs and the retry succeeded in that instance so we are already spending too much time on this, but I feel we can make it flawless so that's why we are moving forward.

> What are the reasons you moved waitForResponse after the user clicks on General tab?

This is exactly how to fix this. I think that the API call was successful, but the `waitForResponse` failed to catch it on time and produced a race condition issue.

I have observed it locally to prove this. [Here is the video with successful API call and failing](https://d.pr/i/7SQXVm) `waitForResponse`. [This is the code added](https://d.pr/i/VJiD0X). With this, we have successfully received API but it was called later in the process so it failed to assert it. If I remove that waitForResponse, the test will be successful showing that my UI based assertion helped to solve the flakiness.

My intention is to illustrate why my change to wait for the notice offers an improvement over the original use of `waitForResponse`. However, **I strongly value collaborative thinking and teamwork**, so I don’t want to focus solely on my approach while excluding others’ contributions. There’s always a way to integrate both perspectives, and I am actively working on achieving that. I would leave notice assertion as it stabilizes the test and verifies important UI change while switching product type.

> As a next step here, I would try to use a Promise to make the check for API response to mitigate the risk of the test failing due to flakiness in this place.

I’m open to exploring other approaches, as there are likely several ways to address this. That said, I’d appreciate it if you could review and test my code as well. I’ve tested it extensively, running it a [hundred times](https://d.pr/i/uDAMmm) with great success.
I’m also keen to hear your thoughts and have you evaluate my findings to ensure we’re aligned on the best solution.

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: OK, one more time.

> [!IMPORTANT]
> TLDR: There is nothing to fix here! The only was forward with this PR is to close it.




From @veljkho's comment, that video is from a failure in #53228. As I already mentioned this in this [comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1864239973) very early in this review, *that PR is still in progress*, and _**that test was expected to fail**_. 
In that run (and in that video) the flow was successful because the call to `/wp-json/wc/v3/products/` was actually successful. The test failed because in that commit the URL to be checked was wrong, it had an extra dot in front of it. Check the error: 
`response.url().includes( './wp-json/wc/v3/products/' )`. This was fixed in the next commit.

> [!IMPORTANT]
> Again, _**that is not a flakiness occurrence and should have been ignored in this investigation**_. All the other (few) failures of this test were caused by other parts of the code. This `waitForResponse` should be left alone.

---

This whole nonsense conversation could have been avoided if:
- a real investigation was actually performed, on all available failures, instead of simply removing some code based only on the last failure without even checking where it happened. Even it that response was the point of failure, removing it is not the fix!
- the feedback received very early would have beed actually read.
- there was no extra noise created by useless AI generated replies.

Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: Please @adimoldovan , the video I was showing [was from this other PR and failure](https://github.com/woocommerce/woocommerce/actions/runs/12118305262/job/33782574373). The video is still there as a failure occurrence, which is the same as what I shared.

We are trying to solve flakiness in this test: `merchant/products/block-editor/create-grouped-product-block-editor.spec.js` as per [the issue ticket](https://github.com/woocommerce/woocommerce/issues/50820).

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: > Please @adimoldovan , the video I was showing [was from this other PR and failure](https://github.com/woocommerce/woocommerce/actions/runs/12118305262/job/33782574373). The video is still there as a failure occurrence, which is the same as what I shared.
> 
> We are trying to solve flakiness in this test: `merchant/products/block-editor/create-grouped-product-block-editor.spec.js` as per [the issue ticket](https://github.com/woocommerce/woocommerce/issues/50820).

Fine, my bad. Taking the step back, do we know why that response failed?

Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: Please see my [observation](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867384010) from above.

It seems to me like a race condition issue where it attempted to wait for a response but was already received and failed to catch it. If we involve it earlier in the process, it will increase its chances of being caught. That's why I modified the code to support that. Please I need your expertise to confirm it, and I tested it literally 100 times locally. Also, for the UI-based assertion, if we could leave it as I really feel that is important for the E2E test (even if that maybe we could leave it without).

Comment (adimoldovan) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: > If we involve it earlier in the process, it will increase its chances of being caught

Agreed. This looks to me like the real fix here.

> Also, for the UI-based assertion, if we could leave it as I really feel that is important for the E2E test (even if that maybe we could leave it without).

That assertion won't help with the fix, but it won't hurt the either. Maybe it creates another potential point of failure in the future, without it being a critical check? I can go either way on this, up to you.

**Review by adimoldovan — APPROVED** [review]
Approving, as per [this comment](https://github.com/woocommerce/woocommerce/pull/53310#discussion_r1867995887) the current code version looks like a good fix.
Thank you @veljkho.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js (hunk 1 above)
Comment (veljkho) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/products/block-editor/create-grouped-product-block-editor.spec.js:82]: > Agreed. This looks to me like the real fix here.

Happy to see this, thanks!

> Maybe it creates another potential point of failure in the future

My concern is that due to the very fast PW and the background process of saving a draft, verifying the notice check before proceeding further will make the test more stable and will add regression coverage that the saving effect should happen. This is especially good for external environments.

**Review by juliaamosova — APPROVED** [review]
@veljkho thank you for your additional comments and reasoning behind your last change of moving the `waitForResponse` earlier in the test. I checked back API calls being made when the product editor to create a new product gets opened and there is indeed a call to get products. Catching that response earlier in the test makes sense to me as well. Approving this PR. I see that there are merge conflicts - please resolve the and we can then merge the PR.

**Comment by juliaamosova** [comment]
>I see that there are merge conflicts - please resolve the and we can then merge the PR.

I don't see conflicts anymore. Merging.
